File: backend/src/models/Quote.js
const mongoose = require('mongoose');

const QuoteItemSchema = new mongoose.Schema({
  code: { type: String, trim: true, default: '' }, // ej. MO01
  description: { type: String, required: true, trim: true },
  quantity: { type: Number, required: true, min: 1, default: 1 },
  unit: { type: String, default: 'Uds.' },
  stockDisp: { type: String, default: '' },
  unitPrice: { type: Number, required: true, min: 0 },
  value: { type: Number, required: true, min: 0 } // quantity * unitPrice
});

const FleetUnitSchema = new mongoose.Schema({
  unitType: { type: String, required: true }, // ej. Bus 12 m, Bus 9 m
  quantity: { type: Number, required: true, min: 1 },
  itemsPerUnit: { type: Number, default: 1 }, // ej. 7 sensores / bus
  totalItems: { type: Number, required: true } // quantity * itemsPerUnit
});

const QuoteSchema = new mongoose.Schema({
  quoteNumber: { type: String, required: true, unique: true, uppercase: true, trim: true }, // ej. DA-2026-018
  templateType: { 
    type: String, 
    enum: ['TALLER_DETALLADO', 'PROYECTO_ESPECIAL'], 
    default: 'TALLER_DETALLADO' 
  },
  status: {
    type: String,
    enum: ['BORRADOR', 'ENVIADA', 'APROBADA', 'EN_TALLER', 'FACTURADA', 'RECHAZADA'],
    default: 'BORRADOR'
  },
  
  // Cliente y Vehículo
  clientId: { type: mongoose.Schema.Types.ObjectId, ref: 'Client' },
  clientName: { type: String, required: true },
  clientDoc: { type: String, default: '' },
  clientAddress: { type: String, default: '' },
  clientPhone: { type: String, default: '' },
  clientEmail: { type: String, default: '' },
  
  vehicleId: { type: mongoose.Schema.Types.ObjectId, ref: 'Vehicle' },
  plate: { type: String, uppercase: true, default: '' }, // Matrícula
  vin: { type: String, uppercase: true, default: '' },
  model: { type: String, default: '' },
  
  // Metadatos
  orderType: { type: String, default: 'Taller de Servicios' }, // Tipo Pedido
  referencePerson: { type: String, default: '' }, // Referencia / Contacto
  advisorName: { type: String, default: 'Ruben Basil' },
  advisorPhone: { type: String, default: '934787006' },
  location: { type: String, default: 'Lima' },
  
  // Fechas
  issueDate: { type: Date, default: Date.now }, // Fecha Alta
  validUntil: { type: Date }, // Fecha Validez
  deliveryTerm: { type: String, default: '' }, // Plazo Entrega
  commissionDays: { type: String, default: '' }, // ej. 4 días
  
  // Contenido de la Cotización
  items: [QuoteItemSchema],
  
  // Campos específicos para PROYECTO_ESPECIAL (Plantilla 1)
  projectObject: { type: String, default: '' }, // 1. OBJETO
  fleetUnits: [FleetUnitSchema], // 2. ALCANCE Y CANTIDAD DE UNIDADES
  executionTime: { type: String, default: '' }, // 4. TIEMPO DE EJECUCIÓN
  includes: [{ type: String }], // 5. INCLUYE
  notIncludes: [{ type: String }], // NO INCLUYE
  commercialConditions: [{ type: String }], // 6. CONDICIONES COMERCIALES
  
  // Totales
  currency: { type: String, default: 'PEN' }, // Soles (S/) o USD
  subtotal: { type: Number, required: true, default: 0 },
  igv: { type: Number, default: 0 },
  applyIgv: { type: Boolean, default: false },
  total: { type: Number, required: true, default: 0 },
  
  // Cuentas Bancarias
  bankAccountsSnapshot: [
    {
      bank: String,
      accountNumber: String,
      interbankAccount: String
    }
  ],
  
  // Observaciones y Notas
  paymentCondition: { type: String, default: 'Condición de pago 07 días despues de realizar el servicio.' },
  notes: { type: String, default: 'Nota: Cotización válida por 15 días hábiles desde su emisión.' },
  
  // Firmas
  clientSignature: { type: String, default: null }, // Base64 PNG
  advisorSignature: { type: String, default: null }, // Base64 PNG
  
  // Archivo PDF
  pdfUrl: { type: String, default: '' }
}, { timestamps: true });

// QuoteSchema.index({ quoteNumber: 1 });
QuoteSchema.index({ plate: 1 });
QuoteSchema.index({ status: 1 });

module.exports = mongoose.model('Quote', QuoteSchema);

File: mobile/lib/models/quote.dart
class CatalogItem {
  final String id;
  final String code;
  final String description;
  final String category;
  final double defaultPrice;

  CatalogItem({
    this.id = '',
    required this.code,
    required this.description,
    this.category = 'MANO_OBRA',
    required this.defaultPrice,
  });

  factory CatalogItem.fromJson(Map<String, dynamic> json) {
    return CatalogItem(
      id: json['_id'] ?? '',
      code: json['code'] ?? '',
      description: json['description'] ?? '',
      category: json['category'] ?? 'MANO_OBRA',
      defaultPrice: (json['defaultPrice'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

class QuoteItem {
  String code;
  String description;
  int quantity;
  double unitPrice;
  double value;

  QuoteItem({
    required this.code,
    required this.description,
    required this.quantity,
    required this.unitPrice,
    required this.value,
  });

  Map<String, dynamic> toJson() => {
    'code': code,
    'description': description,
    'quantity': quantity,
    'unitPrice': unitPrice,
    'value': value,
  };

  factory QuoteItem.fromJson(Map<String, dynamic> json) {
    return QuoteItem(
      code: json['code'] ?? '',
      description: json['description'] ?? '',
      quantity: (json['quantity'] as num?)?.toInt() ?? 1,
      unitPrice: (json['unitPrice'] as num?)?.toDouble() ?? 0.0,
      value: (json['value'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

class Quote {
  final String id;
  final String quoteNumber;
  final String templateType;
  final String status;
  final String clientName;
  final String clientDoc;
  final String clientAddress;
  final String clientPhone;
  final String plate;
  final String vin;
  final String model;
  final double total;
  final List<QuoteItem> items;
  final String? clientSignature;
  final String? pdfUrl;

  Quote({
    required this.id,
    required this.quoteNumber,
    required this.templateType,
    required this.status,
    required this.clientName,
    required this.clientDoc,
    required this.clientAddress,
    required this.clientPhone,
    required this.plate,
    required this.vin,
    required this.model,
    required this.total,
    required this.items,
    this.clientSignature,
    this.pdfUrl,
  });

  factory Quote.fromJson(Map<String, dynamic> json) {
    var rawItems = json['items'] as List? ?? [];
    List<QuoteItem> parsedItems = rawItems.map((i) => QuoteItem.fromJson(i)).toList();

    return Quote(
      id: json['_id'] ?? '',
      quoteNumber: json['quoteNumber'] ?? '',
      templateType: json['templateType'] ?? 'TALLER_DETALLADO',
      status: json['status'] ?? 'BORRADOR',
      clientName: json['clientName'] ?? '',
      clientDoc: json['clientDoc'] ?? '',
      clientAddress: json['clientAddress'] ?? '',
      clientPhone: json['clientPhone'] ?? '',
      plate: json['plate'] ?? '',
      vin: json['vin'] ?? '',
      model: json['model'] ?? '',
      total: (json['total'] as num?)?.toDouble() ?? 0.0,
      items: parsedItems,
      clientSignature: json['clientSignature'],
      pdfUrl: json['pdfUrl'],
    );
  }
}
